libs/vm-modules/include/vm_modules/ml/model/model.hpp

```diff
@@ -112,7 +117,46 @@ class VMModel : public fetch::vm::Object
   ModelConfigPtrType model_config_;
   ModelCategory      model_category_ = ModelCategory::NONE;
 
+  static const std::map<std::string, SupportedLayerType>                 layer_types_;
+  static const std::map<std::string, fetch::ml::details::ActivationType> activations_;
+  static const std::map<std::string, fetch::ml::ops::LossType>           losses_;
+  static const std::map<std::string, fetch::ml::OptimiserType>           optimisers_;
+  static const std::map<std::string, uint8_t>                            model_categories_;
+
   void Init(std::string const &model_category);
+
+  void AddLayerSpecificImpl(SupportedLayerType layer, math::SizeType const &inputs,
+                            math::SizeType const &hidden_nodes);
+  void AddLayerSpecificImpl(SupportedLayerType layer, math::SizeType const &inputs,
+                            math::SizeType const &                   hidden_nodes,
+                            fetch::vm::Ptr<fetch::vm::String> const &activation);
+  void AddLayerSpecificImpl(SupportedLayerType layer, math::SizeType const &inputs,
+                            math::SizeType const &             hidden_nodes,
+                            fetch::ml::details::ActivationType activation);
+  void AddLayerSpecificImpl(SupportedLayerType layer, math::SizeType const &output_channels,
+                            math::SizeType const &input_channels, math::SizeType const &kernel_size,
+                            math::SizeType const &stride_size);
+  void AddLayerSpecificImpl(SupportedLayerType layer, math::SizeType const &output_channels,
+                            math::SizeType const &input_channels, math::SizeType const &kernel_size,
+                            math::SizeType const &                   stride_size,
+                            fetch::vm::Ptr<fetch::vm::String> const &activation);
+  void AddLayerSpecificImpl(SupportedLayerType layer, math::SizeType const &output_channels,
+                            math::SizeType const &input_channels, math::SizeType const &kernel_size,
+                            math::SizeType const &             stride_size,
+                            fetch::ml::details::ActivationType activation);
+
+  inline void AssertLayerTypeMatches(SupportedLayerType                layer,
+                                     std::vector<SupportedLayerType> &&valids) const;
+
+  template <typename T>
+  inline T ParseName(std::string const &name, std::map<std::string, T> const &dict,
+                     std::string const &errmsg) const;
+
+  inline fetch::ml::details::ActivationType DeduceActivation(std::string const &name) const;
```

Comment:
remove the DeduceX functions; the implementations are one line only, and you only use each one twice, and it's extremely clear what the inputs to Parse should be in context

Reply:
Done